operators/pkg/controller/common/annotation/controller_version.go

```diff
@@ -50,3 +55,93 @@ func UpdateControllerVersion(client k8s.Client, obj runtime.Object, version stri
 	log.V(1).Info("updating controller version annotation", "namespace", namespace, "name", name, "kind", obj.GetObjectKind())
 	return client.Update(obj)
 }
+
+// ReconcileCompatibility determines if this controller is compatible with a given resource by examining the controller version annotation
+// controller versions 0.9.0+ cannot reconcile resources created with earlier controllers, so this lets our controller skip those resources until they can be manually recreated
+// if an object does not have an annotation, it will determine if it is a new object or if it has been previously reconciled by an older controller version, as this annotation
+// was not applied by earlier controller versions. it will update the object's annotations indicating it is incompatible if so
+func ReconcileCompatibility(client k8s.Client, obj runtime.Object, selector labels.Selector, controllerVersion string) (bool, error) {
+	accessor := meta.NewAccessor()
+	namespace, err := accessor.Namespace(obj)
+	if err != nil {
+		log.Error(err, "error getting namespace", "kind", obj.GetObjectKind().GroupVersionKind().Kind)
+		return false, err
+	}
+	name, err := accessor.Name(obj)
+	if err != nil {
+		log.Error(err, "error getting name", "namespace", namespace, "kind", obj.GetObjectKind().GroupVersionKind().Kind)
+		return false, err
+	}
+	annotations, err := accessor.Annotations(obj)
+	if err != nil {
+		log.Error(err, "error getting annotations", "namespace", namespace, "name", name, "kind", obj.GetObjectKind().GroupVersionKind().Kind)
+		return false, err
+	}
+
+	annExists := annotations != nil && annotations[ControllerVersionAnnotation] != ""
+
+	// if the annotation does not exist, it might indicate it was reconciled by an older controller version that did not add the version annotation,
+	// in which case it is incompatible with the current controller, or it is a brand new resource that has not been reconciled by any controller yet
+	if !annExists {
+		exist, err := checkExistingResources(client, obj, selector)
+		if err != nil {
+			return false, err
+		}
+		if exist {
+			log.Info("Resource was previously reconciled by incompatible controller version and missing annotation, adding annotation", "controller_version", controllerVersion, "namespace", namespace, "name", name, "kind", obj.GetObjectKind().GroupVersionKind().Kind)
+			err = UpdateControllerVersion(client, obj, "0.8.0-UNKNOWN")
+			return false, err
+		}
+		// no annotation exists and there are no existing resources, so this has not previously been reconciled
+		err = UpdateControllerVersion(client, obj, controllerVersion)
+		return true, err
+	}
+
+	currentVersion, err := version.Parse(annotations[ControllerVersionAnnotation])
+	if err != nil {
+		return false, errors.Wrap(err, "Error parsing current version on resource")
+	}
+	minVersion, err := version.Parse("0.9.0-ALPHA")
```

Comment:
This would probably deserve a constant somewhere.

Reply:
I thought about that but at least right now it is only used in this one function